zerver/actions/custom_profile_fields.py

```diff
@@ -208,3 +214,34 @@ def check_remove_custom_profile_field_value(user_profile: UserProfile, field_id:
         raise JsonableError(_("Field id {id} not found.").format(id=field_id))
     except CustomProfileFieldValue.DoesNotExist:
         pass
+
+
+def validate_user_can_update_custom_profile_data(
+    user_profile: UserProfile,
+    data: List[ProfileDataElementUpdateDict],
+) -> None:
+    with transaction.atomic():
+        for custom_profile_field in data:
+            try:
+                field, created = CustomProfileField.objects.get_or_create(
+                    id=custom_profile_field["id"]
+                )
+                if not field.editable_by_user and not user_profile.is_realm_admin:
+                    raise JsonableError(_("User does not have permission to update the field"))
```

Comment:
I think this check can be done in `validate_user_custom_profile_data` itself, since we already query the database to get `CustomProfileField` in that function. Also we should not create the `CustomProfileField` obejct here if it does not exist, we should just return an error.

Also, if you want to add separate functions, it is better to add them in `zerver/lib/users.py` where other validating functions are present. The files in `actions` are for functions which perform the final task of updating the objects in database and we try/prefer to do all validation before calling the function in `actions` file.

Reply:
I have updated the PR now verification is done in `validate_user_custom_profile_data`.